--- a/ntpq/ntpq.c
+++ b/ntpq/ntpq.c
@@ -32,6 +32,7 @@
 #include "ntp_lineedit.h"
 #include "ntp_debug.h"
 #ifdef OPENSSL
+# include "openssl/crypto.h"
 # include "openssl/evp.h"
 # include "openssl/objects.h"
 # include "openssl/err.h"
@@ -3855,6 +3856,10 @@
 	if (!m || !from || to)
 		return; /* Ignore aliases */
 
+	/* Ignore non-FIPS digests in FIPS mode */
+	if (FIPS_mode() && (EVP_MD_flags(m) & EVP_MD_FLAG_FIPS) == 0)
+		return;
+
 	/* Discard MACs that NTP won't accept. */
 	/* Keep this consistent with keytype_from_text() in ssl_init.c. */
 	if (EVP_MD_size(m) > (MAX_MAC_LEN - sizeof(keyid_t)))
