--- a/libntp/a_md5encrypt.c
+++ b/libntp/a_md5encrypt.c
@@ -245,7 +245,7 @@
 
 /*
  * Calculate the reference id from the address. If it is an IPv4
- * address, use it as is. If it is an IPv6 address, do a md5 on
+ * address, use it as is. If it is an IPv6 address, do a sha1 on
  * it and use the bottom 4 bytes.
  * The result is in network byte order.
  */
@@ -269,7 +269,7 @@
 #   endif
 	/* [Bug 3457] DON'T use plain EVP_DigestInit! It would kill the
 	 * flags! */
-	if (!EVP_DigestInit_ex(ctx, EVP_md5(), NULL)) {
+	if (!EVP_DigestInit_ex(ctx, EVP_sha1(), NULL)) {
 		msyslog(LOG_ERR,
 		    "MD5 init failed");
 		EVP_MD_CTX_free(ctx);	/* pedantic... but safe */
