diff --git a/libntp/a_md5encrypt.c b/libntp/a_md5encrypt.c
index 05c109a..6338f12 100644
--- a/libntp/a_md5encrypt.c
+++ b/libntp/a_md5encrypt.c
@@ -85,7 +85,7 @@ MD5authdecrypt(
 
 /*
  * Calculate the reference id from the address. If it is an IPv4
- * address, use it as is. If it is an IPv6 address, do a md5 on
+ * address, use it as is. If it is an IPv6 address, do a sha1 on
  * it and use the bottom 4 bytes.
  */
 u_int32
@@ -100,7 +100,7 @@ addr2refid(sockaddr_u *addr)
 		return (NSRCADR(addr));
 
 	INIT_SSL();
-	EVP_DigestInit(&ctx, EVP_get_digestbynid(NID_md5));
+	EVP_DigestInit(&ctx, EVP_get_digestbynid(NID_sha1));
 	EVP_DigestUpdate(&ctx, (u_char *)PSOCK_ADDR6(addr),
 	    sizeof(struct in6_addr));
 	EVP_DigestFinal(&ctx, digest, &len);
