diff --git a/ntpd/ntp_crypto.c b/ntpd/ntp_crypto.c
index 38f38a1..8c9fb8e 100644
--- a/ntpd/ntp_crypto.c
+++ b/ntpd/ntp_crypto.c
@@ -116,7 +116,7 @@
  * Global cryptodata in host byte order
  */
 u_int32	crypto_flags = 0x0;	/* status word */
-int	crypto_nid = KEY_TYPE_MD5; /* digest nid */
+int    crypto_nid = NID_sha1; /* digest nid */
 char	*sys_hostname = NULL;
 char	*sys_groupname = NULL;
 static char *host_filename = NULL;	/* host file name */
diff --git a/ntpd/ntp_control.c b/ntpd/ntp_control.c
index a18a4d3..1347651 100644
--- a/ntpd/ntp_control.c
+++ b/ntpd/ntp_control.c
@@ -3594,7 +3594,7 @@ static u_int32 derive_nonce(
 	}
 
 	ctx = EVP_MD_CTX_new();
-	EVP_DigestInit(ctx, EVP_get_digestbynid(NID_md5));
+	EVP_DigestInit(ctx, EVP_get_digestbynid(NID_sha1));
 	EVP_DigestUpdate(ctx, salt, sizeof(salt));
 	EVP_DigestUpdate(ctx, &ts_i, sizeof(ts_i));
 	EVP_DigestUpdate(ctx, &ts_f, sizeof(ts_f));
