--- a/ntpd/ntp_crypto.c
+++ b/ntpd/ntp_crypto.c
@@ -150,7 +150,7 @@
  * Global cryptodata in host byte order
  */
 u_int32	crypto_flags = 0x0;	/* status word */
-int	crypto_nid = KEY_TYPE_MD5; /* digest nid */
+int    crypto_nid = NID_sha1; /* digest nid */
 char	*sys_hostname = NULL;
 char	*sys_groupname = NULL;
 static char *host_filename = NULL;	/* host file name */
--- a/ntpd/ntp_control.c
+++ b/ntpd/ntp_control.c
@@ -3657,9 +3657,9 @@
 #   if defined(OPENSSL) && defined(EVP_MD_CTX_FLAG_NON_FIPS_ALLOW)
 	/* [Bug 3457] set flags and don't kill them again */
 	EVP_MD_CTX_set_flags(ctx, EVP_MD_CTX_FLAG_NON_FIPS_ALLOW);
-	EVP_DigestInit_ex(ctx, EVP_get_digestbynid(NID_md5), NULL);
+	EVP_DigestInit_ex(ctx, EVP_get_digestbynid(NID_sha1), NULL);
 #   else
-	EVP_DigestInit(ctx, EVP_get_digestbynid(NID_md5));
+	EVP_DigestInit(ctx, EVP_get_digestbynid(NID_sha1));
 #   endif
 	EVP_DigestUpdate(ctx, salt, sizeof(salt));
 	EVP_DigestUpdate(ctx, &ts_i, sizeof(ts_i));
