Description:  Fix use-after-free in routing socket code.
Origin:       backport, https://bugs.ntp.org/attachment.cgi?id=883 
Bug:          http://bugs.ntp.org/2224
--- a/ntpd/ntp_io.c
+++ b/ntpd/ntp_io.c
@@ -3534,7 +3534,8 @@
 	int select_count = 0;
 	endpt *ep;
 #if defined(HAS_ROUTING_SOCKET)
-	struct asyncio_reader *asyncio_reader;
+	struct asyncio_reader * asyncio_reader;
+	struct asyncio_reader * next_asyncio_reader;
 #endif
 
 	handler_calls++;
@@ -3637,11 +3638,13 @@
 	asyncio_reader = asyncio_reader_list;
 
 	while (asyncio_reader != NULL) {
+		/* callback may unlink and free asyncio_reader */
+		next_asyncio_reader = asyncio_reader->link;
 		if (FD_ISSET(asyncio_reader->fd, &fds)) {
 			++select_count;
-			(asyncio_reader->receiver)(asyncio_reader);
+			(*asyncio_reader->receiver)(asyncio_reader);
 		}
-		asyncio_reader = asyncio_reader->link;
+		asyncio_reader = next_asyncio_reader;
 	}
 #endif /* HAS_ROUTING_SOCKET */
 	
